Implement a couple attributes for <label> elements

frewsxcv opened the pull request:
<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/8178)

<!-- Reviewable:end -->

frewsxcv committed:
Implement 'control' attribute for <label> elements

frewsxcv committed:
Implement 'htmlFor' attribute for <label> elements

frewsxcv commented:
@nox r?

nox commented:
Reviewed 6 of 6 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlelement.rs, line 307 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8178#-K1Q3RsxGBXZZgSAVFe2)** ([raw file](https://github.com/servo/servo/blob/8e9bd1ed8c6ce19bdf114d42ef93cb4758a516ab/components/script/dom/htmlelement.rs#L307)):</sup>
We should probably make the `type` attribute store an `AttrValue::Atom` value, feel free to file a follow-up or to do it here.
<!-- Reviewable comment -K1Q3RsxGBXZZgSAVFe3:-830950309 -->

---

<sup>**[components/script/dom/htmllabelelement.rs, line 56 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8178#-K1Q4a84a4zBFVseElsO)** ([raw file](https://github.com/servo/servo/blob/8e9bd1ed8c6ce19bdf114d42ef93cb4758a516ab/components/script/dom/htmllabelelement.rs#L56)):</sup>
No need for the type here, I think, because you return it so the type is known.
<!-- Reviewable comment -K1Q4a84a4zBFVseElsP:336741587 -->

---

<sup>**[components/script/dom/htmllabelelement.rs, line 47 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8178#-K1Q4xUhiLTSTukofaUe)** ([raw file](https://github.com/servo/servo/blob/6b0cc839fe69b36ee94002876899e8f028cbdb73/components/script/dom/htmllabelelement.rs#L47)):</sup>
We should make the `for` attribute store an `AttrValue::Atom` too, to ease the comparisons against the actual IDs, especially if outside a document.
<!-- Reviewable comment -K1Q4xUhiLTSTukofaUf:-158015350 -->

---

<sup>**[components/script/dom/htmllabelelement.rs, line 60 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8178#-K1Q4wqRdY0rynXeOKNM)** ([raw file](https://github.com/servo/servo/blob/6b0cc839fe69b36ee94002876899e8f028cbdb73/components/script/dom/htmllabelelement.rs#L60)):</sup>
This is probably wrong in the case the element isn't in a document.
<!-- Reviewable comment -K1Q4wqRdY0rynXeOKNN:-1241143144 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8178)
<!-- Sent from Reviewable.io -->

eefriedman commented on components/script/dom/htmllabelelement.rs:
```diff
@@ -40,6 +42,43 @@ impl HTMLLabelElementMethods for HTMLLabelElement {
+
+        let elem = match document_from_node(self).GetElementById(for_)
+                                                 .and_then(Root::downcast::<HTMLElement>) {
+            Some(elem) => elem,
```
`.iter().filter(|elem| elem.is_labelable_element()).next()` would be a bit shorter... or maybe you have some other idea?  Anyway, the current version seems extremely verbose.

eefriedman commented on components/script/dom/htmllabelelement.rs:
```diff
@@ -40,6 +42,43 @@ impl HTMLLabelElementMethods for HTMLLabelElement {
+    fn GetControl(&self) -> Option<Root<HTMLElement>> {
+        let for_ = self.HtmlFor();
+
+        if for_.is_empty() {
```
This is supposed to be hasAttribute; https://github.com/servo/servo/blob/0fc9410f10c1122a05cd36bdf8b7faff12c39555/tests/wpt/web-platform-tests/html/semantics/forms/the-label-element/label-attributes.html has a test for this.

frewsxcv commented:
<sup>**[components/script/dom/htmlelement.rs, line 307 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8178#-K1Q3RsxGBXZZgSAVFe2)** ([raw file](https://github.com/servo/servo/blob/8e9bd1ed8c6ce19bdf114d42ef93cb4758a516ab/components/script/dom/htmlelement.rs#L307)):</sup>
https://github.com/servo/servo/issues/8180
<!-- Reviewable comment -K1QAPvOPPlJ12clCdp_:2023503413 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8178)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
<sup>**[components/script/dom/htmllabelelement.rs, line 60 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8178#-K1Q4wqRdY0rynXeOKNM)** ([raw file](https://github.com/servo/servo/blob/6b0cc839fe69b36ee94002876899e8f028cbdb73/components/script/dom/htmllabelelement.rs#L60)):</sup>
Suggestions on how to address this?
<!-- Reviewable comment -K1QCDvPI8jgmECRFlEi:-382858933 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8178)
<!-- Sent from Reviewable.io -->

eefriedman commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmllabelelement.rs, line 60 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8178#-K1Q4wqRdY0rynXeOKNM)** ([raw file](https://github.com/servo/servo/blob/6b0cc839fe69b36ee94002876899e8f028cbdb73/components/script/dom/htmllabelelement.rs#L60)):</sup>
`if !self.upcast().is_in_doc() { return None }`?
<!-- Reviewable comment -K1QJSm_88HtuE-r9BeG:-216110679 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8178)
<!-- Sent from Reviewable.io -->

nox commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmllabelelement.rs, line 60 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8178#-K1Q4wqRdY0rynXeOKNM)** ([raw file](https://github.com/servo/servo/blob/6b0cc839fe69b36ee94002876899e8f028cbdb73/components/script/dom/htmllabelelement.rs#L60)):</sup>
Apparently whether the node is in the tree or not is only important for the control, so the current code is correct, aside from the `AttrValue::Atom` issue and the better style suggested by @eefriedman.
<!-- Reviewable comment -K1QNq9smfARbSI0sL-C:-1354340219 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8178)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
<sup>**[components/script/dom/htmllabelelement.rs, line 60 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8178#-K1Q4wqRdY0rynXeOKNM)** ([raw file](https://github.com/servo/servo/blob/6b0cc839fe69b36ee94002876899e8f028cbdb73/components/script/dom/htmllabelelement.rs#L60)):</sup>
@eefriedman, If you want to see our discussion: http://logs.glob.uno/?c=mozilla%23servo
<!-- Reviewable comment -K1QOoqkNRRm308VuJQ8:-1779067663 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8178)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
Comments have been addressed

frewsxcv commented:
One sec, need to update a test expectation

frewsxcv commented:
Done, again.

nox commented:
-S-awaiting-review +S-needs-code-changes 

One last nit.
<!-- Reviewable comment -K1QfLZn6osSeIAwIKiC:-1496052204 -->

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlelement.rs, line 307 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8178#-K1Q3RsxGBXZZgSAVFe2)** ([raw file](https://github.com/servo/servo/blob/8e9bd1ed8c6ce19bdf114d42ef93cb4758a516ab/components/script/dom/htmlelement.rs#L307)):</sup>
Could you file one for `for`?
<!-- Reviewable comment -K1QfHKVKpV_SudSYSrT:-1401307272 -->

---

<sup>**[components/script/dom/htmllabelelement.rs, line 57 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8178#-K1QetYj6uCXpFbYiEH_)** ([raw file](https://github.com/servo/servo/blob/00be26ab6cff99f158d0b7ca22e5c0c7d19e5f80/components/script/dom/htmllabelelement.rs#L57)):</sup>
You can use `get_attribute()` and reuse its value to call `GetElementById()`.
<!-- Reviewable comment -K1QetYj6uCXpFbYiEHa:-835571093 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8178)
<!-- Sent from Reviewable.io -->

eefriedman commented:
> @eefriedman, If you want to see our discussion: http://logs.glob.uno/?c=mozilla%23servo

The WPT test seems to have the opposite interpretation; see https://github.com/servo/servo/blob/0fc9410f10c1122a05cd36bdf8b7faff12c39555/tests/wpt/web-platform-tests/html/semantics/forms/the-label-element/label-attributes.html#L50 .

nox commented:
Cc @Ms2ger who wrote that test.

+I-spec-unclear 
<!-- Reviewable comment -K1QhkloZmq82stwPsJl:-2045008222 -->

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8178)
<!-- Sent from Reviewable.io -->

nox commented:
> Cc @Ms2ger who wrote that test.

I have no idea how I managed to be this wrong. He didn't. I guess it's time for me to go sleep.

Anyway, spec is still unclear, I asked around on #whatwg but there is no one at that moment, will report back when I finally get an answer about this.

nox commented:
OK, I think it all boils down to a missing link in the HTML spec to the correct definition of "in a document" in the [DOM spec](https://dom.spec.whatwg.org/#in-a-document):

> When an element or one of its ancestors is the document element, it is in a document.

Sounds like @eefriedman was right all along. Removing I-spec-unclear.

frewsxcv commented:
> missing link in the HTML spec 

https://github.com/whatwg/html/issues/283

Going to wait for https://github.com/servo/servo/pull/8181 to merge before I force push again so I can utilize it

frewsxcv commented:
I addressed the comments, including changing the `for` attribute to be an `Atom`, but am now getting `panicked at 'Atom not found'` errors when I run the relevant tests. I'm not immediately sure what is causing `for` to get set as a differ `AttrValue`

nox commented:
@frewsxcv That's because that's the first time we implement some virtual method for `HTMLLabelElement`, so we need to add a branch in `vtable_for` for that interface.

-S-awaiting-review +S-needs-code-changes 
<!-- Reviewable comment -K1SsxSgyOvCAbUv9YLJ:72807159 -->

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8178)
<!-- Sent from Reviewable.io -->

Ms2ger commented:
I just filed https://github.com/whatwg/html/issues/285

frewsxcv commented:
> That's because that's the first time we implement some virtual method for HTMLLabelElement, so we need to add a branch in vtable_for for that interface.

Thanks. Fixed in the latest force push. Tests work now.

nox commented:
@Ms2ger Want us to wait for that issue to be fixed? LGTM otherwise.

-S-awaiting-review +S-awaiting-answer 
<!-- Reviewable comment -K1TtjWlpgVaGxNgxxwe:-1298470763 -->

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8178)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
From IRC:

> 10:44 AM <frewsxcv> Ms2ger: https://github.com/servo/servo/pull/8178#issuecomment-150917884 can we get a response here?
> 10:44 AM <crowbot1> PR #8178: Implement a couple attributes for <label> elements - https://github.com/servo/servo/pull/8178
> 10:44 AM <frewsxcv> should hopefully just be a quick yes or no
> 10:47 AM <•Ms2ger> I guess it's fine

@bors-servo r=nox

bors-servo commented:
:pushpin: Commit 98a6fc0 has been approved by `nox`

<!-- @bors-servo r=nox 98a6fc07e2acb44e764158365579f17bbd037ff0 -->

bors-servo commented:
:hourglass: Testing commit 98a6fc0 with merge af8a1cb...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/44), [gonk](http://build.servo.org/builders/gonk/builds/45), [linux-dev](http://build.servo.org/builders/linux-dev/builds/38), [linux-rel](http://build.servo.org/builders/linux-rel/builds/38), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/45), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/45), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/50)
